src/runtime/pkg/kata-monitor/cri.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func dial(ctx context.Context, addr string) (net.Conn, error) {
+	return (&net.Dialer{}).DialContext(ctx, unixProtocol, addr)
+}
+
+func parseEndpointWithFallbackProtocol(endpoint string, fallbackProtocol string) (protocol string, addr string, err error) {
+	if protocol, addr, err = parseEndpoint(endpoint); err != nil && protocol == "" {
+		fallbackEndpoint := fallbackProtocol + "://" + endpoint
+		protocol, addr, err = parseEndpoint(fallbackEndpoint)
+		if err == nil {
+			monitorLog.Warningf("Using %q as endpoint is deprecated, please consider using full url format %q.", endpoint, fallbackEndpoint)
+		}
+	}
+	return
+}
+
+func parseEndpoint(endpoint string) (string, string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", "", err
+	}
+
+	switch u.Scheme {
+	case "tcp":
+		return "tcp", u.Host, nil
+
+	case "unix":
+		return "unix", u.Path, nil
+
+	case "":
+		return "", "", fmt.Errorf("using %q as endpoint is deprecated, please consider using full url format", endpoint)
+
+	default:
+		return u.Scheme, "", fmt.Errorf("protocol %q not supported", u.Scheme)
+	}
+}
+
+// getSandboxes get kata sandbox from the container engine.
+// this will be called only after monitor start.
+func (km *KataMonitor) getSandboxes() (map[string]string, error) {
+
+	sandboxMap := map[string]string{}
+	runtimeClient, runtimeConn, err := getRuntimeClient(km.runtimeEndpoint)
+	if err != nil {
+		return sandboxMap, err
+	}
+	defer closeConnection(runtimeConn)
+
+	filter := &pb.PodSandboxFilter{
+		State: &pb.PodSandboxStateValue{
+			State: pb.PodSandboxState_SANDBOX_READY,
+		},
+	}
+
+	request := &pb.ListPodSandboxRequest{
+		Filter: filter,
+	}
+	monitorLog.Debugf("ListPodSandboxRequest: %v", request)
+	r, err := runtimeClient.ListPodSandbox(context.Background(), request)
+	monitorLog.Debugf("ListPodSandboxResponse: %v", r)
+	if err != nil {
+		return sandboxMap, err
+	}
+
+	for _, pod := range r.Items {
+		request := &pb.PodSandboxStatusRequest{
+			PodSandboxId: pod.Id,
+			Verbose:      true,
+		}
+
+		r, err := runtimeClient.PodSandboxStatus(context.Background(), request)
+		if err != nil {
+			return sandboxMap, err
+		}
+
+		lowRuntime := ""
+		highRuntime := ""
+		var res map[string]interface{}
+		if err := json.Unmarshal([]byte(r.Info["info"]), &res); err != nil {
+			monitorLog.WithError(err).WithField("pod", r).Error("failed to Unmarshal pod info")
+			continue
+		} else {
+			monitorLog.WithField("pod info", res).Debug("")
+			// get high level container runtime
+			pointer, _ := gojsonpointer.NewJsonPointer("/runtimeSpec/annotations/io.container.manager")
+			rt, _, _ := pointer.Get(res)
+			if rt != nil {
+				if str, ok := rt.(string); ok {
+					if str == "cri-o" {
+						highRuntime = RuntimeCRIO
+					} else {
+						highRuntime = RuntimeContainerd
+					}
+				}
+			}
+
+			// get low level container runtime
+			keys := []string{"/runtimeType", "/runtimeSpec/annotations/io.kubernetes.cri-o.RuntimeHandler"}
```

Comment:
This looks pretty crio specific to get the low level runtime. Will this annotaton exist in case of containerd?

Reply:
the first one "/runtimeType" is for containerd. It exists in containerd but not in cri-o.
the second one "/runtimeSpec/[...]/io.kubernetes.cri-o.RuntimeHandler" is for cri-o. It exists in cri-o but not in containerd.

Reply:
We better have literals as constants